docs/2022/puzzles/day07.md

```diff
@@ -1,14 +1,149 @@
 https://adventofcode.com/2022/day/7
 
+## Solution
+
+First of all, we need to come up with a way to parse out input code:
+
+```Scala
+def input (str: String) = str.linesIterator.map {
+    case s"$$ cd $directory" => Cmd(ChangeDirectory(directory))
+    case s"$$ ls" => Cmd(ListFiles)
+    case s"dir $directory" => Directory(directory)
+    case s"$size $file" => File(size.toInt, file)
+}.toList
+```
+We need to create types for commands, to differentiate the input:
+
+```Scala
+enum Command:
+    case ChangeDirectory(directory: String)
+    case ListFiles
+enum TerminalOutput:
+    case Cmd(cmd: Command)
+    case Directory(name: String)
+    case File(size: Int, name: String)
+```
+
+Now let's make a directory structure, in which we will define files as [`mutable.Map`](https://dotty.epfl.ch/api/scala/collection/mutable/Map.html), that can contain name (String) and size (Integer), will have reference to parent directory, and will be able to contain subdirectories:
+
+```Scala
+class DirectoryStructure(val name: String,
+                         val subDirectories: mutable.Map[String, DirectoryStructure],
+                         val files: mutable.Map[String, Int],
+                         val parent: DirectoryStructure | Null)
+```
+We have to come up with a way to calculate directory size -- we can use [`sum`](https://www.scala-lang.org/files/archive/api/current/scala/collection/immutable/List.html#sum[B%3E:A](implicitnum:scala.math.Numeric[B]):B) for the size of all files in directory and define size of all of the following subdirectories recursively, which will take care of problem:
+
+```Scala
+def directorySize(dir: DirectoryStructure): Int =
+    dir.files.values.sum + dir.subDirectories.values.map(directorySize).sum
+```
+
+Now we need to create a function, to transfer input in directory form. For that we can use [`match`](https://docs.scala-lang.org/tour/pattern-matching.html) and separate input, -- for that we can use cases and recursion will do the rest for us:
+
+```Scala
+def buildState(input: List[TerminalOutput], currentDir: DirectoryStructure | Null, rootDir: DirectoryStructure): Unit = input match
+    case Cmd(ChangeDirectory("/")) :: t => buildState(t, rootDir, rootDir)
+    case Cmd(ChangeDirectory("..")) :: t => buildState(t, currentDir.parent, rootDir)
+    case Cmd(ChangeDirectory(name)) :: t => buildState(t, currentDir.subDirectories(name), rootDir)
+    case Cmd(ListFiles) :: t => buildState(t, currentDir, rootDir)
+    case File(size, name) :: t =>
+        currentDir.files.put(name, size)
+        buildState(t, currentDir, rootDir)
+    case Directory(name) :: t =>
+        currentDir.subDirectories.put(name, DirectoryStructure(name, mutable.Map.empty, mutable.Map.empty, currentDir))
+        buildState(t, currentDir, rootDir)
+    case Nil => ()
+```
+
+And now, we need to assemble the program, in part one, we will just search for all directories with size smaller `100000`, and calculate the sum of their sizes. 
```

Comment:
```suggestion
And now, we need to assemble the program, in part one, we will search for all directories with size smaller `100000`, and calculate the sum of their sizes. 
```

Reply:
we should follow the inclusive language guide: https://docs.scala-lang.org/contribute/inclusive-language-guide.html#the-words-easy-simple-quick-and-trivial